// botbuilder-dialogs-adaptive/lib/actions/getConversationMembers.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetConversationMembers = void 0;
const adaptive_expressions_1 = require("adaptive-expressions");
const botbuilder_dialogs_1 = require("botbuilder-dialogs");
/**
 * Calls `BotFrameworkAdapter.getConversationMembers()` and sets the result to a memory property.
 */
class GetConversationMembers extends botbuilder_dialogs_1.Dialog {
    /**
     * Initializes a new instance of the [GetConversationMembers](xref:botbuilder-dialogs-adaptive.GetConversationMembers) class.
     *
     * @param property Property path to put the value in.
     */
    constructor(property) {
        super();
        if (property) {
            this.property = new adaptive_expressions_1.StringExpression(property);
        }
    }
    /**
     * @param property The key of the conditional selector configuration.
     * @returns The converter for the selector configuration.
     */
    getConverter(property) {
        switch (property) {
            case 'property':
                return new adaptive_expressions_1.StringExpressionConverter();
            case 'disabled':
                return new adaptive_expressions_1.BoolExpressionConverter();
            default:
                return super.getConverter(property);
        }
    }
    /**
     * Starts a new [Dialog](xref:botbuilder-dialogs.Dialog) and pushes it onto the dialog stack.
     *
     * @param dc The [DialogContext](xref:botbuilder-dialogs.DialogContext) for the current turn of conversation.
     * @param _options Optional. Initial information to pass to the dialog.
     * @returns A `Promise` representing the asynchronous operation.
     */
    beginDialog(dc, _options) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.disabled && this.disabled.getValue(dc.state)) {
                return yield dc.endDialog();
            }
            const result = yield this.getMembers(dc.context);
            dc.state.setValue(this.property.getValue(dc.state), result);
            return yield dc.endDialog(result);
        });
    }
    getMembers(context) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            const conversationId = (_b = (_a = context.activity) === null || _a === void 0 ? void 0 : _a.conversation) === null || _b === void 0 ? void 0 : _b.id;
            if (!conversationId) {
                throw new Error('[GetConversationMembers]: The getMembers operation needs a valid conversation id.');
            }
            const connectorClient = this.getConnectorClient(context);
            const teamMembers = yield connectorClient.conversations.getConversationMembers(conversationId);
            return teamMembers;
        });
    }
    getConnectorClient(context) {
        var _a;
        const client = context.adapter && 'createConnectorClient' in context.adapter
            ? context.adapter.createConnectorClient(context.activity.serviceUrl)
            : (_a = context.turnState) === null || _a === void 0 ? void 0 : _a.get(context.adapter.ConnectorClientKey);
        if (!client) {
            throw new Error('This method requires a connector client.');
        }
        return client;
    }
    /**
     * @protected
     * Builds the compute Id for the [Dialog](xref:botbuilder-dialogs.Dialog).
     * @returns A `string` representing the compute Id.
     */
    onComputeId() {
        return `GetConversationMembers[${this.property.toString()}]`;
    }
}
exports.GetConversationMembers = GetConversationMembers;
GetConversationMembers.$kind = 'Microsoft.GetConversationMembers';
